pantab/src/pantab.c

```diff
@@ -1,10 +1,47 @@
 #define PY_SSIZE_T_CLEAN
 #include <Python.h>
 
+#include "cffi.h"
+#include "tableauhyperapi.h"
 #include "reader.h"
 #include "writer.h"
 
+
+// Function pointers, initialized by `init` function
+#define C(RET, NAME, ARGS) RET(*NAME)ARGS = NULL;
+HYPERAPI_FUNCTIONS(C)
+#undef C
+
+static PyObject *init(PyObject *Py_UNUSED(dummy), PyObject *args) {
+    bool ok;
+#define C(RET, NAME, ARGS) PyObject* NAME##_arg;
+HYPERAPI_FUNCTIONS(C)
+#undef C
+    const char* formatStr =
+#define C(RET, NAME, ARGS) "O"
+HYPERAPI_FUNCTIONS(C)
+#undef C
+    ;
+    
+    ok = PyArg_ParseTuple(args, formatStr
+#define C(RET, NAME, ARGS) , &NAME##_arg
+HYPERAPI_FUNCTIONS(C)
+#undef C
+    );
+    if (!ok)
+        return NULL;
+
+    // TODO: check that we get an instance of CDataObject; else will segfault
+#define C(RET, NAME, ARGS) NAME = (RET(*)ARGS) (((CDataObject *) NAME##_arg)->c_data);
+HYPERAPI_FUNCTIONS(C)
+#undef C
+
+    Py_RETURN_NONE;
+}
+
 static PyMethodDef methods[] = {
+    {"init", init, METH_VARARGS,
```

Comment:
Rather than this can you just move the initialization logic to `PyInit_libpantab`?

Reply:
Afaik, I cannot pass any parameters from Python to `PyInit_libpantab`. The `init` method needs the CFFI functions, though, so that it can extract the function pointers from it.

Maybe it would be feasible to reimplement the complete "extract function objects from HyperAPI" in C instead of Python. I don't think that would be the right trade-off, though. In general, Python is more readable than in C, and the initialization logic is not performance-critical, so implementing it in C doesn't provide us any perf benefit

Reply:
Hmm OK. Maybe renaming to something like `load_functions` is appropriate then - I just don't want to have confusion in the future of having two init methods in the extension, one being the builtin that gets triggered on actual module initialization and then this one that is called explicitly after the module is initialized

Reply:
renamed to `load_hapi_functions`